### ISSUE TITLE
Registering a new drop zone while dragging an element

### ISSUE THREAD

**Issue by m4nnke** [isaacHagoel/svelte-dnd-action#479]
Hi, 
first of all great work with your library.

I have a use case, where I display the nodes of a simplified abstract syntax tree of a latex document using dropZones as inner Nodes of the graph. It's a LOT of nesting. Inner Nodes can be collapsed so parts of the tree aren't visible. Everything works quite smoothly :).

Now to the point:
While dragging an element, a new dropzone is being mounted (its part of a collapsed inner node that opens after hovering over it). The problem is, that it's not possible to drop the dragged element in the newly created drop zone. (Probably because it is not yet registered in the typeToDropZones Map or something). Is there a way to register a new drop zone while dragging an element?

Thanks in advance

**Comment by isaacHagoel** [comment]
Hi,
Do you mind making a simple REPL that illustrates the issue?
Based on your description I would expect it to work (so maybe there's a bug)

**Comment by m4nnke** [comment]
Thanks for the fast reply. Maybe my description was not detailed enough. I hope this illustrates my problem: 
https://svelte.dev/repl/9af9c4f11c2f4d888438b8b29d9e6c52?version=4.2.0

**Comment by isaacHagoel** [comment]
(cont) the key is to call `observe` on the new zone and add the event listeners. the event listener on the window shouldn't be added again.

**Comment by isaacHagoel** [comment]
@m4nnke i merged that PR but now that i am thinking about it I reckon that the opposite case is probably also an issue...
as in if the user removes a zone mid drag and it unregisters, the lib won't stop observing it and I wonder if that would cause any errors.

**Comment by m4nnke** [comment]
Maybe this is linked with the potential bug I mentioned in #480 first message all the way down (occurs when removing the source zone on drag). You can try it out in the test repo I created. I'll think about that tomorrow. I'm from Europe, it's quiet late here ;)

**Comment by m4nnke** [comment]
I tried my old implementation and it seems, the issue is not occuring there. I stumbled upon this issue too earlier today.

**Comment by m4nnke** [comment]
I think the problem is the fact that rewatching the dragged element leads to reobservation, of the lists, wich leads to a reinitialisation of the variables inside the `observe` function. The next time the dragged element enters a dropzone things go wrong (I'm not sure what exactly). In my first impementation the observe loop keeps running the whole time so the variables stay in tact.
A quick fix in the current implementation: making the 4 variables in the observe function global and cleaning up the variables after a drop has occurred. That seems to fix the issues I encountered. Maybe that gives you an idea what is wrong. I can't tell if it fixes #482.

**Comment by m4nnke** [comment]
The quickfix semi-solves #482, but my old implementation seems to solve #482 too.

**Comment by isaacHagoel** [comment]
@m4nnke 
I reverted the change for now (it broke the crazy nesting example it turns out) and re-opening the ticket after spending a long time re-diving into the relevant parts of the lib and old tickets such as #362 and the fix for it (scheduling the dz for removal after drop rather than removing in real time).
I discovered that in nested scenarios the lower level dropzones (that are items in higher level zones) register and unregister as part of a normal drag operation at the parent level (because the shadow element is created etc). 
Initially I tried adding code to deal with that (which also means doing a better version of the fix to #362 because we need to also be able to remove dynamically "for real" otherwise nested scenarios break) but then some questions started creeping into my mind. If we allow adding and removing zones mid drag and expect the lib to adjust within the same drag operation it raises edge-case questions such as: What should happen if I remove the origin zone and then after a second add "it" back and then drop outside of any zone. Is it still the origin zone?
In other words, I need some more time to think about it.
In the meantime will hiding it with css when it is collapsed be a proper workaround for you?

**Comment by m4nnke** [comment]
Well, thanks for the effort, using css with `display: none` does fix the issue, sadly it slows down the startup, because all the elements need to get in the dom (in large documents there are > 500 nodes = draggable elements and up to 7 nestng levels). The leafs of the tree are mostly iframes with latexjs library inside. `Display:none` seems to break my latexjs inside the iframes. I probably take the first implementation of the issue and use that combined with not being able to close elements while dragging. That seems to work for me.

**Comment by isaacHagoel** [comment]
I see. Alright. I'll keep you posted

**Comment by isaacHagoel** [comment]
Just a short update. I am working on a PR that would address some of the core issues that makes this difficult. Specifically it adds much tighter control over nested zones life-cycle and eliminates the noise they currently create by registering and being destroyed as the parent element is dragged. You can watch #483 to stay in the loop about that.
After that's merge I will re-examine supporting dynamically added and removed zones.
Notes to self:
- Don't forget to consider the aria aspects of this (announce that a new zone is available?)
- Consider a concept of zoneID to deal with same zone disappearing and reappearing, especially when it is the origin zone.
